### PR TITLE
feat(IO.Config): consider XDG specification

### LINKED ISSUE #28: Use XDG Specification
Hey, you should respect the XDG specification: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html 

I'll try to implement that soonish, but wanted to create an issue for that first.

Good article about that: https://0x46.net/thoughts/2019/02/01/dotfile-madness/

### PR DESCRIPTION
**Pull request by mstruebing**
I've got it to work :tada: 

Please keep in mind that this is my first serious contribution to a Haskell project and I'm not that experienced with it to be honest. Maybe there is something that can be improved about my programming style, if so please let me know. :)

closes #28

### CODE REVIEW

File: src/IO/Config.hs [hunk 1]
```diff
@@ -39,7 +40,19 @@ defaultConfig = Config {
 }
 
 getDir :: IO FilePath
-getDir = (++ "/.taskell") <$> getHomeDirectory
+getDir =
```
Comment (isovector) [inline src/IO/Config.hs]: Hi! Someone posted this on reddit asking for a code review, so here I am!

I'd put this entire function into an IO block: 

```haskell
getDir = do
```

File: src/IO/Config.hs [hunk 2]
```diff
@@ -39,7 +40,19 @@ defaultConfig = Config {
...
+    let
```
Comment (isovector) [inline src/IO/Config.hs]: Then get rid of all of these bindings except for `directoryName`. As a general principle, you don't want to `let`-define monadic values that you are only going to use once.

File: src/IO/Config.hs [hunk 3]
```diff
@@ -39,7 +40,19 @@ defaultConfig = Config {
...
+        oldConfigDir = (++ "/." ++ directoryName) <$> getHomeDirectory
+        xdgDefaultConfigDirectory = (++ "/.config/" ++ directoryName) <$> getHomeDirectory
+        iOMaybeXdgEnvironmentConfigDirectory = lookupEnv "XDG_CONFIG_HOME"
+        xdgDirectory = do
+            maybeXdgEnvironmentConfigDirectory <- iOMaybeXdgEnvironmentConfigDirectory
+            case maybeXdgEnvironmentConfigDirectory  of
+                 Just xdgEnvironmentConfigDirectory -> pure $ xdgEnvironmentConfigDirectory ++ "/" ++ directoryName
+                 Nothing -> xdgDefaultConfigDirectory
+    in
+    oldConfigDir >>= doesDirectoryExist >>= (\oldconfigAvailable -> if oldconfigAvailable then oldConfigDir else xdgDirectory)
```
Comment (isovector) [inline src/IO/Config.hs]: Now that you are in the `do` block, you don't need to use `>>=` anymore. This becomes the more readable:

```
  home <-getHomeDirectory
  let oldConfigPath = home ++ "'/." ++ directoryName
  oldConfigAvailable <- doesDirectoryExist oldConfigPath

  if oldConfigAvailable
    then pure oldConfigPath
    else do
      xdgEnv = lookupEnv "XDG_CONFIG_HOME"
      pure $ case xdgEnv  of
        Just xdgConfigHome -> xdgConfigHome ++ "/" ++ directoryName
        Nothing -> home ++ "/.config/" ++ directoryName
```

**Comment by smallhadroncollider** [comment]
Thanks @isovector, I was going to respond with something similar! But on a deadline today, so double thanks.

File: src/IO/Config.hs (hunk 3 above)
Comment (mstruebing) [reply src/IO/Config.hs]: Thank you, I've somehow thought that `do` blocks wouldn't be considered good practice but it seems I'm wrong. I will provide a fix soonish, thanks.

Comment (smallhadroncollider) [reply src/IO/Config.hs]: Some people don't like `do` blocks as they hide what's really going on, but they're definitely preferable to using `let` to store monadic values. You could remove the `do` completely by breaking it up into smaller functions, but probably not necessary.

Comment (mstruebing) [reply src/IO/Config.hs]: Currently try to refactor into smaller funcs :)

**Comment by mstruebing** [comment]
So I've split it into smaller functions. If you like the `do`-block more I can remove my last commit.

**Comment by smallhadroncollider** [comment]
That's looking much tidier!

One last thing: rather than using `if` in a lambda after `>>=` you could use [`bool`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bool.html#v:bool) to avoid the lambda:

```
oldConfigPath >>= doesDirectoryExist >>= bool xdgConfigPath oldConfigPath
```

Note that `bool` takes the `False` value *first*. I think there are a couple of places you could do this.

**Comment by smallhadroncollider** [comment]
It's probably also better to keep `lookupEnv "XDG_CONFIG_HOME"` as a `Maybe` and then use a `case` expression in `xdgConfigPath` instead. Then you don't have to return empty strings or convert `Maybe`s into `Bool` at all. (Then the `bool` point above would only apply in one case)

**Comment by smallhadroncollider** [comment]
In fact you could just do:

```
xdgConfigPath :: IO FilePath
xdgConfigPath = fromMaybe xdgDefaultConfig <$> lookupEnv "XDG_CONFIG_HOME"
```

Which would replace three lines 51-62.

Haskell's fun to refactor!

**Comment by smallhadroncollider** [comment]
Should probably be using `filepath`s [`</>`](https://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:-60--47--62-) operator to join paths together too. But that's my bad! Only learnt about it last week.

**Comment by mstruebing** [comment]
Hey, I've done some of your suggestions.
But I didn't got this function:

```hs
xdgConfigPath :: IO FilePath
xdgConfigPath = fromMaybe xdgDefaultConfig <$> lookupEnv "XDG_CONFIG_HOME"
```

to work in any variation I've tried. :/

**Comment by smallhadroncollider** [comment]
Give me a minute and I'll give it a go. My `~/.stack` directory got massive so I wiped it yesterday, but means I need to do a re-build, which takes a while!

**Comment by smallhadroncollider** [comment]
Ah, yeah, sorry, forgot it was in an `IO` too.

```
xdgConfigPath :: IO FilePath
xdgConfigPath = lookupEnv "XDG_CONFIG_HOME" >>= maybe xdgDefaultConfig return
```

Not quite as tidy. I'm sure there's probably some way to do it without the `return`, but my brain is a bit slow at the moment.

**Comment by smallhadroncollider** [comment]
Ah, it's a bloom' applicative isn't it:

```
xdgConfigPath :: IO FilePath
xdgConfigPath = fromMaybe <$> xdgDefaultConfig <*> lookupEnv "XDG_CONFIG_HOME"
```

I'm so bad at spotting those.

**Comment by mstruebing** [comment]
Many thanks! :))
I feel like I understood all these weird operators right now. :tada:

**Comment by smallhadroncollider** [comment]
The code looks really tidy now. I'll just need to check that we're not missing anything and then I'll pull it in. Probably won't get round to that today because the aforementioned deadline that I'm currently procrastinating around by writing Haskell! But hopefully by end of day Monday.

Thanks for your efforts and well done for getting Haskell to do anything on your first attempt. Took me about five years.

**Comment by smallhadroncollider** [comment]
Gah! Ok, changing the base branch isn't a good idea. Sorry, made this page very messy.

**Pull request merged**